### ISSUE TITLE
[Bug]: FP8 KV cache causes systematic decode/prefill logprob mismatch on Hopper FA3

### ISSUE THREAD

**Issue by my0901** [vllm-project/vllm#54035]
## Your current environment

```text
OS: Ubuntu 22.04.5 LTS (x86_64), kernel 5.15.0-119-generic
Python: 3.11.15
vLLM: 0.28.0 (also reproduced on 0.27.1)
PyTorch: 2.13.0+cu130
CUDA runtime: 13.0
NVIDIA driver: 580.159.03
GPU: NVIDIA H100 80GB HBM3
Model: Qwen3-4B (BF16 weights)
Attention backend: FLASH_ATTN (FlashAttention 3)
KV cache dtype: fp8 / fp8_e4m3; control: auto (BF16)
Prefix caching: disabled
Execution mode: eager
End-to-end comparison: VLLM_BATCH_INVARIANT=1 (set before importing vLLM)
```

### Summary

With FP8 KV cache, vLLM returns systematically different logprobs depending on whether a token is scored through decode (token-by-token generation) or through a full-prefill rescore (`prompt_logprobs` on the same sequence). On Qwen3-4B (H100, vLLM 0.28.0, greedy, 224 generated tokens, `VLLM_BATCH_INVARIANT=1`), the first mismatch is at absolute sequence position **97**, with 115–117 mismatching generated-token positions and max `|Δlogprob|` ≈ 0.56. The root cause is in the bundled FlashAttention 3 Hopper FP8 kernel (`vllm-project/flash-attention`): decode and prefill select different KV tile widths (`kBlockN=96` vs `192` at `head_dim=128`), and a direct FA3 kernel reproducer below shows the same boundary at KV length `seqlen_k=97`.

### End-to-end reproduction

1. Greedy-generate 224 tokens with `logprobs=1`; record each output token's logprob.
2. Concatenate prompt + output; run a full-prefill rescore with `prompt_logprobs`.
3. Compare logprob at each generated-token position.

Set `VLLM_BATCH_INVARIANT=1` before importing vLLM.

**Observed (unpatched, vLLM 0.28.0, FP8 KV cache):**

| Prompt | First mismatch position | Mismatching positions | Max `\|Δlogprob\|` |
| --- | ---: | ---: | ---: |
| `tell me a story` | 97 | 115 / 224 | 0.50 |
| `讲个故事` | 97 | 117 / 224 | 0.40 |

Without `VLLM_BATCH_INVARIANT=1`, decode and one-shot prefill can differ even with BF16 because of unrelated batch-shape effects. Those differences must not be used alone to attribute a residual to this FA3 FP8 kernel bug.

### Minimal kernel reproduction

The test calls `vllm.vllm_flash_attn.flash_attn_varlen_func` twice with the same FP8 K/V tensors and the same final query row: once as a full causal query (prefill), and once as a one-row query (decode).

```python
import torch
from vllm.vllm_flash_attn import flash_attn_varlen_func

device = "cuda"
nheads_q, nheads_kv, head_dim = 12, 2, 128

def run(seqlen):
    torch.manual_seed(42)
    k = torch.randn(seqlen, nheads_kv, head_dim, device=device).to(torch.float8_e4m3fn)
    v = torch.randn(seqlen, nheads_kv, head_dim, device=device).to(torch.float8_e4m3fn)
    q = torch.randn(seqlen, nheads_q, head_dim, device=device).to(torch.float8_e4m3fn)
    descale = torch.ones(1, nheads_kv, device=device)

    def call(query):
        q_len = query.shape[0]
        return flash_attn_varlen_func(
            q=query, k=k, v=v,
            cu_seqlens_q=torch.tensor([0, q_len], dtype=torch.int32, device=device),
            cu_seqlens_k=torch.tensor([0, seqlen], dtype=torch.int32, device=device),
            max_seqlen_q=q_len, max_seqlen_k=seqlen,
            softmax_scale=head_dim**-0.5, causal=True,
            k_descale=descale, v_descale=descale, q_descale=descale,
            block_table=None, fa_version=3,
        )

    prefill = call(q)[-1].float()
    decode = call(q[-1:])[0].float()
    return torch.equal(prefill, decode), (prefill - decode).abs().max().item()

for length in list(range(88, 112)) + [128, 160, 192, 256]:
    print(length, run(length))
```

**Observed kernel result (unpatched FA3, H100, head_dim=128):**

The sweep covers 28 lengths: 88–111, plus 128, 160, 192, and 256.

`97` below is a **KV-cache sequence length**, not an E2E token position. `19/28` means 19 of these 28 independently tested lengths have a different final-row attention output.

| KV dtype | First mismatching length | Mismatching sampled lengths | Difference |
| --- | ---: | ---: | ---: |
| FP8 E4M3 | **97** | **19/28** | about `3.9e-3` at 97; maximum about `9.8e-3` |
| BF16 control | 160 | 3/28 | maximum about `2e-3` |

Example unpatched FP8 output:

```text
seqlen_cache= 96: max_diff=0.00000000 bitwise_equal=True
seqlen_cache= 97: max_diff=0.00390625 bitwise_equal=False  <-- MISMATCH
...
[RESULT] FP8 (e4m3fn): first mismatch at seqlen=97; 19 of 28 tested lengths differ
```

The E2E first-mismatch position (97) aligns with `kBlockN_decode(96) + 1`.

### Root cause

`use_one_mma_wg` is enabled for a sufficiently small effective query length:

- [`hopper/heuristics.h#L10-L14`](https://github.com/vllm-project/flash-attention/blob/main/hopper/heuristics.h#L10-L14)

The FP8 branch of [`hopper/tile_size.h`](https://github.com/vllm-project/flash-attention/blob/main/hopper/tile_size.h) then selects different K tile widths:

- decode `kBlockN=96` for `head_dim=128`: [`#L55-L61`](https://github.com/vllm-project/flash-attention/blob/main/hopper/tile_size.h#L55-L61)
- prefill `kBlockN=192` (two-level accumulation enabled): [`#L66-L73`](https://github.com/vllm-project/flash-attention/blob/main/hopper/tile_size.h#L66-L73)

vLLM routes both decode and prefill through `flash_attn_varlen_func` via the FLASH_ATTN backend, so this fork tiling choice is what the E2E comparison exercises.

Different K-tile boundaries change the online-softmax reduction and FP8 accumulation/scaling path, producing the boundary at length 97.

Related: [sgl-project/sglang#25790](https://github.com/sgl-project/sglang/issues/25790) (same root cause in sibling `sgl-attn` fork).

### Expected behavior

For the same Q/K/V values, the final-row attention output should not depend on whether the query is supplied through the decode-shaped or prefill-shaped call. In batch-invariant vLLM mode, the corresponding end-to-end decode/rescore logprobs should also agree.

### Proposed fix

Fix belongs in **`vllm-project/flash-attention`** (`hopper/tile_size.h`): retain decode's small `kBlockM=64`, but make decode use the same FP8 `kBlockN` as prefill. vLLM main would then bump its flash-attention pin after that fork PR merges.

### Before submitting a new issue

- [x] I searched open and closed vLLM issues for an existing report of this specific Hopper FA3 FP8 decode/prefill tiling mismatch. No exact duplicate found; related FP8 KV issues (e.g. scaling/LSE/backend bugs) describe different root causes.
- [x] I asked the chatbot on the vLLM documentation site as required by the bug-report form.
- [x] Added the relevant runtime, GPU, model, and serving configuration, and checked the report for tokens, credentials, private paths, and confidential data.

**Comment by my0901** [comment]
Fix PR in the flash-attention fork: https://github.com/vllm-project/flash-attention/pull/184

**Comment by CarrotSwordsman** [comment]
Full reproduction on a second sm_90 card (H20 96GB, driver 535, vllm 0.28.0+cu129 wheel, torch 2.13.0+cu129) — every boundary matches your H100 numbers:

**Kernel-level reproducer** (unmodified, same script):

- first mismatch exactly at `seqlen=97`, `|Δ| = 3.9e-3`
- 19/28 tested lengths differ, max `|Δ| ≈ 1.03e-2` (at seqlen 192) — vs your 9.8e-3

**E2E** (Qwen3-4B, greedy, 224 tokens, `VLLM_BATCH_INVARIANT=1`, FA3 confirmed active):

| KV dtype | first mismatch (abs. position) | mismatches / 224 | max \|Δ\| |
|---|---:|---:|---:|
| FP8 E4M3 | **97** | 91 | 0.059 |
| BF16 control | none | **0** | **0** |

The BF16 control being bit-identical between decode and rescore pins the entire residual on the FP8 KV path, and the E2E first-mismatch position aligning with the kernel boundary (`kBlockN_decode(96) + 1`) matches your analysis. (Our absolute mismatch count / magnitude are smaller than yours — 91/224 and 0.059 vs ~115/224 and ~0.5 — likely different generation content — but "systematic divergence starting at position 97" reproduces exactly.)

So this is not card-specific: same behavior on H100 and H20, i.e. purely the sm_90 FA3 FP8 decode/prefill tiling choice in `hopper/tile_size.h` as diagnosed. Happy to help validate a `kBlockN` alignment patch on this H20 if you open the flash-attention fork PR.
